ee/clickhouse/views/person.py

```diff
@@ -1,30 +1,93 @@
-from typing import Callable, Dict, Optional, Tuple
+import json
+from typing import Any, Callable, Dict, List, Optional, Tuple, cast
 
+from django.db.models.query import Prefetch, QuerySet
 from rest_framework.decorators import action
 from rest_framework.exceptions import NotFound
 from rest_framework.request import Request
 from rest_framework.response import Response
 
 from ee.clickhouse.client import sync_execute
 from ee.clickhouse.models.person import delete_person
+from ee.clickhouse.models.property import parse_prop_clauses
 from ee.clickhouse.queries.clickhouse_retention import ClickhouseRetention
 from ee.clickhouse.queries.clickhouse_stickiness import ClickhouseStickiness
 from ee.clickhouse.queries.funnels import ClickhouseFunnelPersons, ClickhouseFunnelTrendsPersons
 from ee.clickhouse.queries.trends.lifecycle import ClickhouseLifecycle
 from ee.clickhouse.sql.person import GET_PERSON_PROPERTIES_COUNT
 from posthog.api.person import PersonViewSet
-from posthog.api.utils import format_next_url, format_offset_absolute_url
-from posthog.constants import INSIGHT_FUNNELS, FunnelVizType
+from posthog.api.utils import format_offset_absolute_url
+from posthog.constants import FunnelVizType
 from posthog.decorators import cached_function
+from posthog.exceptions import UnsupportedFeature
 from posthog.models import Event, Filter, Person
 
+BASE_FILTER_PERSONS_QUERY = """
+SELECT person.id FROM person
+LEFT OUTER JOIN (
+    SELECT * FROM person_distinct_id WHERE person_distinct_id.team_id = %(team_id)s
+) person_distinct_id
+ON person.id = person_distinct_id.person_id
+WHERE team_id = %(team_id)s"""
+
+
+def filter_persons_ch(team_id: int, request: Request, queryset: QuerySet) -> QuerySet:
```

Comment:
Doing all of this in the view is really sketchy.

Given that this endpoint exists in both clickhouse and postgres, the proper approach is to extract a query object and have a test_factory testing both.

Reply:
What do you mean by `extract a query object and have a test_factory testing both`?

Reply:
In other words it should behave as e.g. trends does. There's an endpoint which creates a Query object and all the work is done in there. There's also a test_factory which creates tests for the feature and is used for both postgres and ee tests.

Reply:
I still don't quite get what this'd look like in a cleaner way.

Reply:
See refactoring done in https://github.com/PostHog/posthog/pull/5842 - I separated out a query class, same way e.g. trends and everything else works.